reference.c

```diff
@@ -0,0 +1,591 @@
+        if (isdigit(*MD)) {
+            // match
+            int len = strtol((char *)MD, (char **)&MD, 10);
+            while (iseq < b->core.l_qseq && len) {
+                if ((next_op = next_cigar_op(cigar, &ncigar, cig_skip,
+                                             &iseq, &cig_ind, &cig_op,
+                                             &cig_len)) < 0)
+                    return -1;
+
+                if (next_op != BAM_CMATCH &&
+                    next_op != BAM_CEQUAL) {
+                    print_error("MD2ref",
+                                "MD:Z and CIGAR are incompatible");
+                    return -1;
+                }
+
+                if (iref < ref_len)
+                    ref[iref] = seq_nt16_str[bam_seqi(seq, iseq)];
+                iseq++;
+                iref++;
+                len--;
+            }
+        } else if (*MD == '^') {
+            // deletion
+            MD++;
+            while (*MD && isalpha(*MD)) {
+                if ((next_op = next_cigar_op(cigar, &ncigar, cig_skip,
+                                             &iseq, &cig_ind, &cig_op,
+                                             &cig_len)) < 0)
+                    return -1;
+
+                if (next_op != BAM_CDEL) {
+                    print_error("MD2ref",
+                                "MD:Z and CIGAR are incompatible");
+                    return -1;
+                }
+
+                if (iref < ref_len)
+                    ref[iref] = *MD;
+
+                MD++;
+                iref++;
+            }
+        } else {
+            // substitution
+            if ((next_op = next_cigar_op(cigar, &ncigar, cig_skip,
+                                         &iseq, &cig_ind, &cig_op,
+                                         &cig_len)) < 0)
+                return -1;
+
+            if (next_op != BAM_CMATCH && next_op != BAM_CDIFF) {
+                print_error("MD2ref", "MD:Z and CIGAR are incompatible");
+                return -1;
+            }
+            if (iref < ref_len)
+                ref[iref] = *MD;
+
+            MD++;
+            iref++;
+            iseq++;
+        }
+    }
+
+    return 1;
+}
+
+static int MD2ref(samFile *in, sam_hdr_t *h, hts_idx_t *idx, char *reg,
+                  FILE *outfp, int verbose) {
+    bam1_t *b = bam_init1();
+    int r, last_tid = -99;
+    size_t ref_len = 0;
+    char *ref = NULL;
+    int ret = -1;
+
+    hts_itr_t *iter = NULL;
+    if (idx && reg) {
+        iter = sam_itr_querys(idx, h, reg);
+        if (!iter) {
+            print_error("reference", "failed to parse region '%s'", reg);
+            goto err;
+        }
+    }
+
+    while ((r = iter
+                ? sam_itr_next(in, iter, b)
+                : sam_read1(in, h, b)) >= 0) {
+        // check b->core.tid and flush old seq.
+        if (b->core.tid != last_tid) {
+            if (last_tid >= 0)
+                dump_ref(h, iter, last_tid, ref, ref_len, outfp, 1);
```

Comment:
Should dump_ref verbose always be 1 here?  The same question for the other dump_refs in this function.